File: initialize/KHI_periodic_lowres/perturb.m
function perturb(cfg, xg)
% perturb plasma from initial_conditions file

narginchk(2,2)
validateattributes(cfg, {'struct'}, {'scalar'})
validateattributes(xg, {'struct'}, {'scalar'})
%% READ IN THE SIMULATION INFORMATION

x1=xg.x1(3:end-2);
x2=xg.x2(3:end-2);
x3=xg.x3(3:end-2);
lx1=xg.lx(1); lx2=xg.lx(2); lx3=xg.lx(3);


%% LOAD THE FRAME OF THE SIMULATION THAT WE WANT TO PERTURB
dat = loadframe3Dcurvnoelec(cfg.indat_file);
lsp = size(dat.ns,4);

%% SCALE EQ PROFILES UP TO SENSIBLE BACKGROUND CONDITIONS
scalefact=2.75;
nsscale=zeros(size(dat.ns));
for isp=1:lsp-1
    nsscale(:,:,:,isp) = scalefact * dat.ns(:,:,:,isp);
end %for
nsscale(:,:,:,lsp) = sum(nsscale(:,:,:,1:6),4);   %enforce quasineutrality


%% APPLY THE PERTURBATION
v0=500;                   % background flow value, actually this will be turned into a shear in the Efield input file
densfact=10;              % factor by which the density increases over the shear region - see Keskinen, et al (1988)
voffset=2*v0/densfact;    % plays the role of the neutral wind from K88; prevents singularity - this form results from tweaking asymptotic value of ne to give a certain density jump
ell=1e3;                  % scale length for shear transition

nsperturb=zeros(size(dat.ns));
for isp=1:lsp
  for ix2=1:xg.lx(2)
    amplitude=randn(xg.lx(1),1,xg.lx(3));    %AGWN, note this can make density go negative so error checking needed below
    amplitude=0.01*amplitude;

    nsperturb(:,ix2,:,isp)=nsscale(:,ix2,:,isp).*(2*v0+voffset)./(-v0*tanh((x2(ix2))/ell)+v0+voffset);

    nsperturb(:,ix2,:,isp)=nsperturb(:,ix2,:,isp)+amplitude.*nsscale(:,ix2,:,isp);        %add some noise to seed instability
  end %for
end %for
nsperturb=max(nsperturb,1e4);                        %enforce a density floor (particularly need to pull out negative densities which can occur when noise is applied)
nsperturb(:,:,:,lsp)=sum(nsperturb(:,:,:,1:6),4);    %enforce quasineutrality


%% KILL OFF THE E-REGION WHICH WILL DAMP THE INSTABILITY (AND USUALLY ISN'T PRESENT IN PATCHES)
x1ref=150e3;     %where to start tapering down the density in altitude
dx1=10e3;
taper=1/2+1/2*tanh((x1-x1ref)/dx1);
for isp=1:lsp-1
   for ix3=1:xg.lx(3)
       for ix2=1:xg.lx(2)
           nsperturb(:,ix2,ix3,isp)=1e6+nsperturb(:,ix2,ix3,isp).*taper;
       end %for
   end %for
end %for
nsperturb(:,:,:,lsp)=sum(nsperturb(:,:,:,1:6),4);    %enforce quasineutrality


%% WRITE OUT THE RESULTS TO A NEW FILE
ymd = dat.simdate(1:3);
UTsec = dat.simdate(4)*3600;
writedata(ymd, UTsec, nsperturb, dat.vs1, dat.Ts, cfg.outdir, cfg.file_format);

end % function

File: init/GDI_Eregion/GDI_round/perturb.py
from __future__ import annotations
import typing as T
import numpy as np
import numpy.random

import gemini3d.read
import gemini3d.write


def perturb(cfg: dict[str, T.Any], xg: dict[str, T.Any]):
    """
    perturb plasma from initial_conditions file
    """

    # %% READ IN THE SIMULATION INFORMATION
    # trim ghost cells
    x1 = xg["x1"][2:-2]
    x2 = xg["x2"][2:-2]
    x3 = xg["x3"][2:-2]

    # %% LOAD THE FRAME OF THE SIMULATION THAT WE WANT TO PERTURB
    dat = gemini3d.read.frame(cfg["indat_file"], var=["ns", "Ts", "vs1"])
    ns = dat["ns"]
    lsp = ns.shape[0]

    # %% Choose a single profile from the center of the eq domain as a reference
    ix2 = xg["lx"][1] // 2
    ix3 = xg["lx"][2] // 2

    nsscale = np.zeros_like(ns)
    for i in range(lsp):
        nprof = ns[i, :, ix2, ix3]
        for j in range(xg["lx"][1]):
            for k in range(xg["lx"][2]):
                nsscale[i, :, j, k] = nprof

    # %% SCALE EQ PROFILES UP TO SENSIBLE BACKGROUND CONDITIONS
    scalefact = 2.75 * 6 / 8
    for i in range(lsp - 1):
        nsscale[i, :, :, :] = scalefact * nsscale[i, :, :, :]
    nsscale[-1, :, :, :] = nsscale[:-1, :, :, :].sum(axis=0)
    # enforce quasineutrality

    # %% GDI EXAMPLE (PERIODIC) INITIAL DENSITY STRUCTURE AND SEEDING
    ellx2 = 50e3  # gradient scale length for patch/blob
    ellx3 = 50e3
    x2ctr = 150e3
    # x21 = -150e3  # location on one of the patch edges
    # x22 = 150e3  # other patch edge
    # ell = 20e3
    nepatchfact = 3  # density increase factor over background

    nsperturb = np.zeros_like(ns)
    for i in range(lsp - 1):
        for j in range(xg["lx"][1]):
            for k in range(xg["lx"][2]):
                amplitude = numpy.random.standard_normal(xg["lx"][0])
                # AWGN - note that can result in subtractive effects on density so apply a floor later!!!
                amplitude = 0.01 * amplitude
                # amplitude standard dev. is scaled to be 1% of reference profile

                ## original data, infinite patch
                # nsperturb[i, :, j, k] = nsscale[i, :, j, k] + nepatchfact * nsscale[i, :, j, k] * (
                #    1 / 2 * np.dict((x2[j] - x21) / ell) - 1 / 2 * np.dict((x2[j] - x22) / ell)
                # )

                # Circular patch
                nsperturb[i, :, j, k] = nsscale[i, :, j, k] + nepatchfact * nsscale[i, :, j, k] * (
                    np.exp(-((np.sqrt((x2[j] - x2ctr) ** 2 + x3[k] ** 2)) ** 8) / 2 / ellx2**8)
                )
                # patch, note offset in the x2 index!!!!

                if (j > 9) and (j < xg["lx"][1] - 10):
                    # do not apply noise near the edge (avoids boundary artifacts)
                    nsperturb[i, :, j, k] = nsperturb[i, :, j, k] + amplitude * nsscale[i, :, j, k]

    nsperturb = np.maximum(nsperturb, 1e4)
    # enforce a density floor (particularly need to pull out negative densities
    # which can occur when noise is applied)
    nsperturb[-1, :, :, :] = nsperturb[:-1, :, :, :].sum(axis=0)  # enforce quasineutrality

    # %% KILL OFF THE E-REGION WHICH WILL DAMP THE INSTABILITY (AND USUALLY ISN'T PRESENT IN PATCHES)
    x1ref = 200e3
    # where to start tapering down the density in altitude
    dx1 = 10e3
    taper = 0.5 + 0.5 * np.dict((x1 - x1ref) / dx1)
    for i in range(lsp - 1):
        for ix3 in range(xg["lx"][2]):
            for ix2 in range(xg["lx"][1]):
                nsperturb[i, :, ix2, ix3] = 1e6 + nsperturb[i, :, ix2, ix3] * taper

    nsperturb[-1, :, :, :] = nsperturb[:-1, :, :, :].sum(axis=0)
    # enforce quasineutrality

    # %% WRITE OUT THE RESULTS TO the same file
    gemini3d.write.state(
        cfg["indat_file"],
        dat,
        ns=nsperturb,
        # file_format=cfg["file_format"],
    )
